platform/x11/os_x11.cpp

```diff
@@ -2499,48 +2499,66 @@ static String get_mountpoint(const String &p_path) {
 }
 
 Error OS_X11::move_to_trash(const String &p_path) {
-	String trashcan = "";
+	String trash_can = "";
 	String mnt = get_mountpoint(p_path);
 
+	// If there is a directory "[Mountpuint]/.Trash-[UID]/files", use it as the trash can.
 	if (mnt != "") {
 		String path(mnt + "/.Trash-" + itos(getuid()) + "/files");
 		struct stat s;
 		if (!stat(path.utf8().get_data(), &s)) {
-			trashcan = path;
+			trash_can = path;
 		}
 	}
 
-	if (trashcan == "") {
+	// Otherwise, if ${XDG_DATA_HOME} is defined, use "${XDG_DATA_HOME}/Trash/files" as the trash can.
+	if (trash_can == "") {
 		char *dhome = getenv("XDG_DATA_HOME");
 		if (dhome) {
-			trashcan = String(dhome) + "/Trash/files";
+			trash_can = String(dhome) + "/Trash/files";
 		}
 	}
 
-	if (trashcan == "") {
+	// Otherwise, if ${HOME} is defined, use "${HOME}/.local/share/Trash/files" as the trash can.
+	if (trash_can == "") {
 		char *home = getenv("HOME");
 		if (home) {
-			trashcan = String(home) + "/.local/share/Trash/files";
+			trash_can = String(home) + "/.local/share/Trash/files";
 		}
 	}
 
-	if (trashcan == "") {
-		ERR_PRINTS("move_to_trash: Could not determine trashcan location");
+	// Issue an error if none of the previous locations is appropriate for the trash can.
+	if (trash_can == "") {
+		ERR_PRINTS("move_to_trash: Could not determine the trash can location");
 		return FAILED;
 	}
 
+	// Create needed directories for decided trash can location.
 	List<String> args;
 	args.push_back("-p");
-	args.push_back(trashcan);
-	Error err = execute("mkdir", args, true);
-	if (err == OK) {
-		List<String> args2;
-		args2.push_back(p_path);
-		args2.push_back(trashcan);
-		err = execute("mv", args2, true);
+	args.push_back(trash_can);
+	int retval;
+	Error err = execute("mkdir", args, true, NULL, NULL, &retval);
```

Comment:
I wonder why this code isn't using `DirAccess::make_dir_recursive()` instead of calling the system `mkdir` command? Same for `DirAccess:rename()` vs `mv`.

Reply:
I don't know why it wasn't used until now, like I said, I am new here :D
I have tried to implement it now, but I have problems, either I am using those functions wrong, or they don't work properly for some reason.

I've replaced this part(execution of `mkdir`) with:
```
      // Create needed directories for decided trash can location.
      DirAccess *dir_access = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
      Error err = dir_access->make_dir_recursive(trash_can);
  
      // Issue an error if trash can is not created proprely.
      if (err != OK) {
          ERR_PRINTS("move_to_trash: Could not create the trash can \"" + trash_can + "\"");
          return err;
      }
```

and it works as expected. 
But replacing the execution of `mv` call with:
```
        // The trash can is successfully created, now move the given resource to it.
	err = dir_access->rename(p_path, trash_can);

	// Issue an error if mv failed to move the given resource to the trash can.
	if (err != OK) {
		ERR_PRINTS("move_to_trash: Could not move the resource \"" + p_path + "\" to the trash can \"" + trash_can + "\"");
		return FAILED;
	}
```

doesn't work even in normal cases. Asset was not removed, nor is the error reported.

Looking at the definition of the method `DirAccessUnix::rename` in `drivers/unix/dir_access_unix.cpp`, I actually can't understand what is happening in the last line:
``` return ::rename(p_path.utf8().get_data(), p_new_path.utf8().get_data()) == 0 ? OK : FAILED;```
Which `rename` function is this?
Even using "step" in gdb just moved over it, like `::rename` is not even a function.

At my current state, I will either need help with understanding what is this `::rename` function to check why it doesn't work properly or I can leave the second part as the execution of the `mv` command.

Reply:
The `::` in `::rename` is used to access the outer scope, and thus not call `DirAccessUnix::rename` recursively but instead call the stdio `rename` method: http://www.cplusplus.com/reference/cstdio/rename/

Reply:
Oh, I didn't know there is `rename` in `stdio.h`, thank you very much!
I'll try to create minimal C program to check with exact strings and find out what is the problem.

Reply:
I've found the problem. `rename` from `stdio.h` can not move/rename files across multiple mounting points, which is the case in my setup.

Therefore, I would suggest that execution of `mv` stays like it was and execution of `mkdir` can be replaced with `DirAccess::make_dir_recursive()`.

If you agree and have no additional suggestions, I'll update the PR.

Reply:
Sounds good to me for now, please add a comment before the `mv` call explaining your findings.

Reply:
Hmm, ideally move_to_trash should never really move data across mountpoints at all. That smells like a bug to me.

Reply:
As far as I'm concerned, the option named `Delete` that doesn't actually delete the asset but moves it to the other location is misleading for me.
After this is done I would like to try to implement a global setting for default trash can which can be empty and if that setting is empty, the option will be named `Delete` like it is now, but if that setting has some path, the option would be named `Move to Trash`.
Of course, this will take me a lot of time to implement, so I wanted to improve the current version first. I would also like to hear your opinions on this, before I start with the implementation.

@akien-mga
Sure, no problem.

@hpvb
I agree, but the current algorithm suggests thrash can locations in this order:
1) "[Mountpoint of deleted asset]/.Trash-[UID]/files"
2) "${XDG_DATA_HOME}/Trash/files"
3) "${HOME}/.local/share/Trash/files"

In my current setup, Godot is on my SSD in "${HOME}/", but new projects on which I play and learn Godot are on my HDD in "/hdd/projects/".
When Godot works with current algorithm, it desides to use location 3. as my trash can and I don't like that.

I understand that it can't pick the ideal location that will fit every user and that is why I would like to try to implement the global setting that everyone can pick for themselves.
But even so, users will be able to pick the trash can on the other mountpoints which will result in error once they use `Move to Trash`.